Revert breaking type changes.

markstory opened the pull request:
As per #12335 this reverts breaking changes made in #12286

markstory committed:
Revert breaking type changes.

As per #12335 this reverts breaking changes made in #12286

dereuromark commented on src/Filesystem/File.php:
```diff
@@ -450,29 +450,29 @@ public function exists(): bool
      * Returns the "chmod" (permissions) of the file.
      *
-     * @return string|null Permissions for the file, or null in case of an error
+     * @return string|bool Permissions for the file, or false in case of an error
```
bool must be false here
Same for the other cases, as true can never be returned (we only abuse false here for null).

markstory replied:
PHPStan doesn't seem to have blown up so far. Why would `bool` cause issues?

markstory replied:
Ok. We use `bool` quite extensively as it is a type while `false` is not. If tooling works better with `string|false` that sounds like a separate change.

markstory commented:
@dereuromark Yes I think we can schedule replacing the bool|other changes for 5.0. Another option would be to replace the 'Filesystem' package with an improved implementation avoiding the need to break compatibility in a subtle way.

markstory committed:
Convert bool hints to false where possible in Filesystem

We don't ever return true, and hinting to bool presents the potential to
do so. Using a narrower type gives users less options to consider.
